test/spmd/tensor/test_megatron_example.py

```diff
@@ -96,18 +115,57 @@ def aggregate_output(outputs):
     return dist_mod
 
 
+def shard_self_attn(m, device_type):
+    pg = dist.distributed_c10d._get_default_group()
+    start_idx = 0
+    device_mesh = DeviceMesh(
+        device_type,
+        list(range(start_idx, start_idx + pg.size())),
+        dim_groups=[pg],
+    )
+    col_wise_sharding = [Shard(0)]
+    row_wise_sharding = [Shard(1)]
+    replicate = [Replicate()]
+    m.qkv.weight = torch.nn.Parameter(
+        distribute_tensor(m.qkv.weight, device_mesh, col_wise_sharding)
+    )
+    m.proj.weight = torch.nn.Parameter(
+        distribute_tensor(m.proj.weight, device_mesh, row_wise_sharding)
+    )
+    if m.qkv.bias is not None:
+        m.qkv.bias = torch.nn.Parameter(
+            distribute_tensor(m.qkv.bias, device_mesh, col_wise_sharding)
+        )
+    if m.proj.bias is not None:
+        m.proj.bias = torch.nn.Parameter(
+            distribute_tensor(m.proj.bias, device_mesh, replicate)
+        )
+    m = _replicate_input_tensor(m, device_mesh, replicate)
+    m.proj = _aggregate_local_tensor(m.proj)
+    m.qkv.weight.register_hook(functools.partial(_gradient_hook, m.qkv.weight))
```

Comment:
Why do we need to do this for the TP case? I don't think I set a grad hook for any of my examples.

Reply:
This is indeed needed because it's for 2D parallel. The way FSDP was designed that they will use gradients from local tensors and we need to pass gradients from DT to local tensors. cc: @wanchaol